openhands/memory/condenser.py

```diff
@@ -1,24 +1,412 @@
+            )
+        if keep_first < 0:
+            raise ValueError(f'keep_first ({keep_first}) cannot be negative')
+        if max_size < 1:
+            raise ValueError(f'max_size ({keep_first}) cannot be non-positive')
+
+        self.max_size = max_size
+        self.keep_first = keep_first
+        self.llm = llm
+
+        super().__init__()
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """If the history is too long, use an LLM to select the most important events."""
+        if len(events) <= self.max_size:
+            return events
+
+        target_size = self.max_size // 2
+        head = events[: self.keep_first]
+
+        events_from_tail = target_size - len(head)
+
+        message: str = """You will be given a list of actions, observations, and thoughts from a coding agent.
+        Each item in the list has an identifier. Please sort the identifiers in order of how important the
+        contents of the item are for the next step of the coding agent's task, from most important to least
+        important."""
+
+        response = self.llm.completion(
+            messages=[
+                {'content': message, 'role': 'user'},
+                *[
+                    {
+                        'content': f'<ID>{e.id}</ID>\n<CONTENT>{e.message}</CONTENT>',
+                        'role': 'user',
+                    }
+                    for e in events
+                ],
+            ],
+            response_format={
+                'type': 'json_schema',
+                'json_schema': {
+                    'name': 'ImportantEventSelection',
+                    'schema': ImportantEventSelection.model_json_schema(),
+                },
+            },
+            # response_format=ImportantEventSelection,
+        )
+
+        response_ids = ImportantEventSelection.model_validate_json(
+            response.choices[0].message.content
+        ).ids
+
+        self.add_metadata('all_event_ids', [event.id for event in events])
+        self.add_metadata('response_ids', response_ids)
+
+        # Filter out any IDs from the head and trim the results down
+        head_ids = [event.id for event in head]
+        response_ids = [
+            response_id for response_id in response_ids if response_id not in head_ids
+        ][:events_from_tail]
+
+        # If the response IDs aren't _long_ enough, iterate backwards through the events and add any unfound IDs to the list.
+        for event in reversed(events):
+            if len(response_ids) >= events_from_tail:
+                break
+            if event.id not in response_ids:
+                response_ids.append(event.id)
+
+        # Grab the events associated with the response IDs
+        tail = [event for event in events if event.id in response_ids]
 
-            # TODO If the llm fails with ContextWindowExceededError, we can try to condense the memory chunk by chunk
-            raise
+        return head + tail
```

Comment:
I am a bit unsure where, if at all, does the code account for the "pair" that is used with tool use?

Currently, with tool events, the codebase assumes that the observation and the action that triggered it should be "linked" together. I mean, if an action has no obs, it will linger in the agent as "pending" and may wait forever for its obs. If the obs has no action, I think the agent used to give an error (not sure if it still does, but litellm would, because the API would, if it gets there).

IMHO we should _somewhere_ make sure that the agent history is as it should be. I'm not sure if here is the best place. Maybe not, just want to pick your brain on this. Am I missing something? In your experiments, have you seen a litellm error complaining about a tool result but there was no tool id, or the other way around? Is the history resulted from condensers unaware of whether the actions and obs (with tools) are still "paired"?

Reply:
Condensers do not account for the paired events from tool use. I can see an argument for making them keep pairs together, but I didn't explore that direction because it was more complex and the simpler approach yielded results. We might want to add that structural constraint later, but I think there's still value and information in split pairs (maybe because `Observation` content usually summarizes the action taken anyways).

> IMHO we should _somewhere_ make sure that the agent history is as it should be.

I fully agree with this. The `history` field in the `State` objects managed by agent controllers should be this place, and condensers were designed to leave that unchanged while giving the agents a more compressed "view".

> In your experiments, have you seen a litellm error complaining about a tool result but there was no tool id, or the other way around?

I've seen no such errors. My guess is that these are all handled by the agent controller, which sees the full history. The condensed view is only ever seen by the agent's `step` method, which converts events to strings anyways and just uses them for completion calls.

Reply:
> there's still value and information in split pairs (maybe because Observation content usually summarizes the action taken anyways).

If the observation is sent without its tool call, we can get this kind of error
<img width="814" alt="image" src="https://github.com/user-attachments/assets/f664bd00-5e6a-4c99-b2cb-9d31f602f8d0" />

So I don't think it happens? For example, [this code](https://github.com/All-Hands-AI/OpenHands/blob/157ff4a4b905e12f973a3ac4d2053a65212fe031/openhands/agenthub/codeact_agent/codeact_agent.py#L336) doesn't return the observation content, instead the content is filled if it is ["complete"](https://github.com/All-Hands-AI/OpenHands/blob/157ff4a4b905e12f973a3ac4d2053a65212fe031/openhands/agenthub/codeact_agent/codeact_agent.py#L474) (in other words, they're all "paired")


> My guess is that these are all handled by the agent controller, which sees the full history.

It doesn't though, the only thing that the agent controller does on this, is precisely "pairing" them, but only when we truncate events because of ContextWindowExceeded:
https://github.com/All-Hands-AI/OpenHands/blob/157ff4a4b905e12f973a3ac4d2053a65212fe031/openhands/controller/agent_controller.py#L885

I wonder if what happens (can't check right now) is that the extra actions/obs exist in the agent, but are dropped silently when it sends the `completion` call.

Reply:
I see what you mean. I've no clue why I haven't seen those errors -- even went back through all my benchmark logs and couldn't find any evidence. I need to work through the `Event` -> `Message` pipeline and figure out if there's some danger there.

Reply:
I think I have a sense of what is going on: the `_get_messages` method of the agent is only ever emitting tool call info if it can find the "paired" event. So if the condenser drops the action but not the observation we won't even construct the observation message.

I tested this in situ by cobbling together a condenser that drops all `Action` events and got no errors.

I've also added a unit test (`test_mismatched_tool_call_events`) to the end of `test_codeact_agent.py` that should fail if this behavior ever changes.